File: Napakalaki1/src/napakalaki1/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki1;

import java.util.ArrayList;

/**
 *
 * @author TOSHIBA
 */

public class SpecificBadConsequence extends BadConsequence{
    private ArrayList<TreasureKind> specificVisibleTreasures;
    private ArrayList<TreasureKind> specificHiddenTreasures;
    
    public SpecificBadConsequence(String t, int l, ArrayList<TreasureKind> tV, 
                            ArrayList<TreasureKind> tH){
        super(t, l);
        specificVisibleTreasures = new ArrayList(tV);
        specificHiddenTreasures = new ArrayList(tH);
    }
    @Override
    public boolean isEmpty() {
        return specificVisibleTreasures.isEmpty() && specificHiddenTreasures.isEmpty();
    }
    @Override
    public void substractVisibleTreasure(Treasure t) {
        int num = specificVisibleTreasures.size();
        for (int i=0; i<num; i++){
            if (t.getType() == specificVisibleTreasures.get(i)){
                specificVisibleTreasures.remove(i);
                break;
            }
        }
    }
    @Override
    public void substractHiddenTreasure(Treasure t) {
        int num = specificHiddenTreasures.size();
        for (int i=0; i<num; i++){
            if (t.getType() == specificHiddenTreasures.get(i)){
                specificHiddenTreasures.remove(i);
                break;
            }
        }
    }
    @Override
    public SpecificBadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h) {
        int size;
        ArrayList<TreasureKind> visible = new ArrayList(specificVisibleTreasures);
        ArrayList<TreasureKind> aux =new ArrayList(visible);
        for (Treasure treasure: v){
            size = aux.size();
            for (int i=0; i<size; i++){
                if (aux.get(i)== treasure.getType()){
                    aux.remove(i);
                    break;
                }
            }
        }       
        for (TreasureKind tKind : aux){
            visible.remove(tKind);
        }
        ArrayList<TreasureKind> hidden = new ArrayList(specificHiddenTreasures);
        aux = new ArrayList(hidden);
        for (Treasure treasure : v){
            size = aux.size();
            for (int i=0; i<size; i++){
                if (aux.get(i) ==treasure.getType()){
                    aux.remove(i);
                    break;
                }
            }
        }  
        for (TreasureKind tKind : aux){
            hidden.remove(tKind);
        }
        SpecificBadConsequence newBadCons= new SpecificBadConsequence("", 0, visible, hidden);
        return newBadCons;
    }
    @Override
    public String toString(){
        return super.toString();
    }
}

File: NapakalakiRuby1/lib/specific_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'bad_consequence.rb'

module NapakalakiGame
  class SpecificBadConsequence < BadConsequence
    attr_reader :specificVisibleTreasures, :specificHiddenTreasures
    def initialize (t, l, sV, sH)
      super(t, l)
      @specificVisibleTreasures = Array.new(sV)
      @specificHiddenTreasures = Array.new(sH)
    end
    public_class_method :new
    def isEmpty
      return (@specificVisibleTreasures.empty?) && (@specificHiddenTreasures.empty?)
    end
    def substractVisibleTreasure (t)
      for i in 0..@specificVisibleTreasures.length-1 do
        if t.type == @specificVisibleTreasures[i]
          @specificVisibleTreasures.delete_at(i)
          break
        end
      end
    end
    def substractHiddenTreasure (t)
      for i in 0..@specificHiddenTreasures.length-1 do
        if t.type == @specificHiddenTreasures[i]
          @specificHiddenTreasures.delete_at(i)
          break
        end
      end
    end
    def adjustToFitTreasureLists (v, h)
      visible = Array.new(@specificVisibleTreasures)
      aux = Array.new(visible)
      v.each do |t|
        for i in 0..aux.length-1 do
          if aux[i] == t.type
            aux.delete_at(i)
            break
          end
        end
      end
      aux.each do |tK|
        #visible.delete_at(visible.find_index(tK))
        visible.delete(tK)
      end
      hidden = Array.new(@specificHiddenTreasures)
      aux = Array.new(hidden)
      h.each do |t|
        for i in 0..aux.length-1 do
          if aux[i] == t.type
            aux.delete_at(i)
            break
          end
        end
      end
      aux.each do |tK|
        #hidden.delete_at(hidden.find_index(tK))
        hidden.delete(tK)
      end
      bC = SpecificBadConsequence.new("", 0, visible, hidden)
    end
    def to_s
      super + "Tesoros Visibles = " + @specificVisibleTreasures.to_s + "\nTesoros Ocultos = " + @specificHiddenTreasures.to_s + "\n"
    end
  end
end
